File: syscalls/ServiceInstaller.cs
using System.Runtime.InteropServices;
using System.ServiceProcess;
using System.Text;

namespace NtServiceInstaller;

public static class ServiceInstaller
{
    public static bool InstallService(string exePath, string serviceName, string? displayName = null, string? description = null)
    {
        IntPtr servicesKeyHandle = IntPtr.Zero;
        IntPtr serviceKeyHandle = IntPtr.Zero;
        
        try
        {
            if (!File.Exists(exePath))
            {
                Console.WriteLine($"Error: Executable file not found: {exePath}");
                return false;
            }
            
            displayName ??= serviceName;
            
            // Open HKLM\SYSTEM\CurrentControlSet\Services
            var servicesPath = NtHelper.InitUnicodeString(NtConstants.SERVICES_KEY);
            var servicesOa = NtHelper.InitObjectAttributes(ref servicesPath);
            
            uint status = NtApi.NtOpenKey(
                out servicesKeyHandle,
                NtConstants.KEY_CREATE_SUB_KEY,
                ref servicesOa);
                
            NtHelper.FreeUnicodeString(ref servicesPath);
            
            if (!NtHelper.NT_SUCCESS(status))
            {
                Console.WriteLine($"Failed to open Services key: 0x{status:X8}");
                return false;
            }
            
            // Create service subkey
            var serviceKeyPath = NtHelper.InitUnicodeString(serviceName);
            var serviceOa = NtHelper.InitObjectAttributes(ref serviceKeyPath, rootDirectory: servicesKeyHandle);
            
            status = NtApi.NtCreateKey(
                out serviceKeyHandle,
                NtConstants.KEY_ALL_ACCESS,
                ref serviceOa,
                0,
                IntPtr.Zero,
                0,
                out uint disposition);
                
            NtHelper.FreeUnicodeString(ref serviceKeyPath);
            
            if (!NtHelper.NT_SUCCESS(status))
            {
                Console.WriteLine($"Failed to create service key: 0x{status:X8}");
                return false;
            }
            
            Console.WriteLine($"Service key created (disposition: {disposition})");
            
            // Set Type = SERVICE_WIN32_OWN_PROCESS
            if (!SetRegistryDWord(serviceKeyHandle, "Type", NtConstants.SERVICE_WIN32_OWN_PROCESS))
                return false;
            
            // Set Start = SERVICE_AUTO_START
            if (!SetRegistryDWord(serviceKeyHandle, "Start", NtConstants.SERVICE_AUTO_START))
                return false;
            
            // Set ErrorControl = SERVICE_ERROR_NORMAL
            if (!SetRegistryDWord(serviceKeyHandle, "ErrorControl", NtConstants.SERVICE_ERROR_NORMAL))
                return false;
            
            // Set ImagePath
            if (!SetRegistryString(serviceKeyHandle, "ImagePath", exePath))
                return false;
            
            // Set DisplayName
            if (!SetRegistryString(serviceKeyHandle, "DisplayName", displayName))
                return false;
            
            // Set ObjectName = LocalSystem
            if (!SetRegistryString(serviceKeyHandle, "ObjectName", "LocalSystem"))
                return false;
            
            // Set Description if provided
            if (!string.IsNullOrEmpty(description))
            {
                if (!SetRegistryString(serviceKeyHandle, "Description", description))
                    Console.WriteLine("Warning: Failed to set description");
            }
            
            Console.WriteLine($"Service '{serviceName}' installed successfully via NT syscalls");
            Console.WriteLine("Note: Service may require system restart or manual SCM refresh to appear");
            
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error installing service: {ex.Message}");
            return false;
        }
        finally
        {
            if (serviceKeyHandle != IntPtr.Zero)
                NtApi.NtClose(serviceKeyHandle);
            if (servicesKeyHandle != IntPtr.Zero)
                NtApi.NtClose(servicesKeyHandle);
        }
    }
    
    public static bool UninstallService(string serviceName)
    {
        IntPtr servicesKeyHandle = IntPtr.Zero;
        IntPtr serviceKeyHandle = IntPtr.Zero;
        
        try
        {
            // Stop service first if running
            try
            {
                using var sc = new ServiceController(serviceName);
                if (sc.Status != ServiceControllerStatus.Stopped)
                {
                    Console.WriteLine($"Stopping service '{serviceName}'...");
                    sc.Stop();
                    sc.WaitForStatus(ServiceControllerStatus.Stopped, TimeSpan.FromSeconds(10));
                }
            }
            catch
            {
                // Service may not exist in SCM yet
            }
            
            // Open Services key
            var servicesPath = NtHelper.InitUnicodeString(NtConstants.SERVICES_KEY);
            var servicesOa = NtHelper.InitObjectAttributes(ref servicesPath);
            
            uint status = NtApi.NtOpenKey(
                out servicesKeyHandle,
                NtConstants.KEY_ENUMERATE_SUB_KEYS,
                ref servicesOa);
                
            NtHelper.FreeUnicodeString(ref servicesPath);
            
            if (!NtHelper.NT_SUCCESS(status))
            {
                Console.WriteLine($"Failed to open Services key: 0x{status:X8}");
                return false;
            }
            
            // Open service subkey
            var serviceKeyPath = NtHelper.InitUnicodeString(serviceName);
            var serviceOa = NtHelper.InitObjectAttributes(ref serviceKeyPath, rootDirectory: servicesKeyHandle);
            
            status = NtApi.NtOpenKey(
                out serviceKeyHandle,
                NtConstants.DELETE,
                ref serviceOa);
                
            NtHelper.FreeUnicodeString(ref serviceKeyPath);
            
            if (!NtHelper.NT_SUCCESS(status))
            {
                if (status == NtConstants.STATUS_OBJECT_NAME_NOT_FOUND)
                {
                    Console.WriteLine($"Service '{serviceName}' does not exist");
                }
                else
                {
                    Console.WriteLine($"Failed to open service key: 0x{status:X8}");
                }
                return false;
            }
            
            // Delete the key
            status = NtApi.NtDeleteKey(serviceKeyHandle);
            
            if (!NtHelper.NT_SUCCESS(status))
            {
                Console.WriteLine($"Failed to delete service key: 0x{status:X8}");
                return false;
            }
            
            Console.WriteLine($"Service '{serviceName}' uninstalled successfully via NT syscalls");
            
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error uninstalling service: {ex.Message}");
            return false;
        }
        finally
        {
            if (serviceKeyHandle != IntPtr.Zero)
                NtApi.NtClose(serviceKeyHandle);
            if (servicesKeyHandle != IntPtr.Zero)
                NtApi.NtClose(servicesKeyHandle);
        }
    }
    
    public static bool StartService(string serviceName)
    {
        try
        {
            using var service = new ServiceController(serviceName);
            
            if (service.Status == ServiceControllerStatus.Running)
            {
                Console.WriteLine($"Service '{serviceName}' is already running");
                return true;
            }
            
            Console.WriteLine($"Starting service '{serviceName}'...");
            service.Start();
            service.WaitForStatus(ServiceControllerStatus.Running, TimeSpan.FromSeconds(30));
            
            Console.WriteLine($"Service '{serviceName}' started successfully");
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error starting service: {ex.Message}");
            return false;
        }
    }
    
    public static bool StopService(string serviceName)
    {
        try
        {
            using var service = new ServiceController(serviceName);
            
            if (service.Status == ServiceControllerStatus.Stopped)
            {
                Console.WriteLine($"Service '{serviceName}' is already stopped");
                return true;
            }
            
            Console.WriteLine($"Stopping service '{serviceName}'...");
            service.Stop();
            service.WaitForStatus(ServiceControllerStatus.Stopped, TimeSpan.FromSeconds(30));
            
            Console.WriteLine($"Service '{serviceName}' stopped successfully");
            return true;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error stopping service: {ex.Message}");
            return false;
        }
    }
    
    public static void GetServiceStatus(string serviceName)
    {
        try
        {
            using var service = new ServiceController(serviceName);
            
            Console.WriteLine($"Service Name: {service.ServiceName}");
            Console.WriteLine($"Display Name: {service.DisplayName}");
            Console.WriteLine($"Status: {service.Status}");
            Console.WriteLine($"Start Type: {service.StartType}");
        }
        catch (InvalidOperationException)
        {
            Console.WriteLine($"Service '{serviceName}' does not exist in SCM");
            Console.WriteLine("Note: Service may exist in registry but not yet loaded by SCM");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error getting service status: {ex.Message}");
        }
    }
    
    private const uint DELETE = 0x00010000;
    
    private static bool SetRegistryDWord(IntPtr keyHandle, string valueName, uint value)
    {
        var valueNameUs = NtHelper.InitUnicodeString(valueName);
        
        try
        {
            IntPtr dataPtr = Marshal.AllocHGlobal(4);
            try
            {
                Marshal.WriteInt32(dataPtr, (int)value);
                
                uint status = NtApi.NtSetValueKey(
                    keyHandle,
                    ref valueNameUs,
                    0,
                    NtConstants.REG_DWORD,
                    dataPtr,
                    4);
                
                if (!NtHelper.NT_SUCCESS(status))
                {
                    Console.WriteLine($"Failed to set {valueName}: 0x{status:X8}");
                    return false;
                }
                
                return true;
            }
            finally
            {
                Marshal.FreeHGlobal(dataPtr);
            }
        }
        finally
        {
            NtHelper.FreeUnicodeString(ref valueNameUs);
        }
    }
    
    private static bool SetRegistryString(IntPtr keyHandle, string valueName, string value)
    {
        var valueNameUs = NtHelper.InitUnicodeString(valueName);
        
        try
        {
            var bytes = Encoding.Unicode.GetBytes(value + "\0");
            IntPtr dataPtr = Marshal.AllocHGlobal(bytes.Length);
            
            try
            {
                Marshal.Copy(bytes, 0, dataPtr, bytes.Length);
                
                uint status = NtApi.NtSetValueKey(
                    keyHandle,
                    ref valueNameUs,
                    0,
                    NtConstants.REG_SZ,
                    dataPtr,
                    (uint)bytes.Length);
                
                if (!NtHelper.NT_SUCCESS(status))
                {
                    Console.WriteLine($"Failed to set {valueName}: 0x{status:X8}");
                    return false;
                }
                
                return true;
            }
            finally
            {
                Marshal.FreeHGlobal(dataPtr);
            }
        }
        finally
        {
            NtHelper.FreeUnicodeString(ref valueNameUs);
        }
    }
}

File: syscalls-cpp/service_installer.cpp
#include "service_installer.h"
#include "nt_api.h"
#include <stdio.h>
#include <wchar.h>

// Helper to set registry DWORD value
static BOOL SetRegistryDWord(HANDLE keyHandle, LPCWSTR valueName, DWORD value) {
    UNICODE_STRING valueNameUs;
    InitUnicodeString(&valueNameUs, valueName);
    
    NTSTATUS status = NtSetValueKey(
        keyHandle,
        &valueNameUs,
        0,
        REG_DWORD,
        &value,
        sizeof(DWORD)
    );
    
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to set %s: 0x%X\n", valueName, status);
        return FALSE;
    }
    return TRUE;
}

// Helper to set registry string value
static BOOL SetRegistryString(HANDLE keyHandle, LPCWSTR valueName, LPCWSTR value) {
    UNICODE_STRING valueNameUs;
    InitUnicodeString(&valueNameUs, valueName);
    
    SIZE_T length = (wcslen(value) + 1) * sizeof(WCHAR);
    
    NTSTATUS status = NtSetValueKey(
        keyHandle,
        &valueNameUs,
        0,
        REG_SZ,
        (PVOID)value,
        (ULONG)length
    );
    
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to set %s: 0x%X\n", valueName, status);
        return FALSE;
    }
    return TRUE;
}

BOOL InstallService(LPCWSTR exePath, LPCWSTR serviceName, LPCWSTR displayName, LPCWSTR description) {
    HANDLE servicesKey = NULL;
    HANDLE serviceKey = NULL;
    NTSTATUS status;
    ULONG disposition;
    BOOL success = FALSE;
    
    if (!InitNtFunctions()) {
        wprintf(L"Failed to initialize NT functions\n");
        return FALSE;
    }
    
    // Open Services registry key
    UNICODE_STRING servicesPath;
    InitUnicodeString(&servicesPath, SERVICES_KEY_PATH);
    
    OBJECT_ATTRIBUTES servicesOa;
    InitObjectAttributes(&servicesOa, &servicesPath, OBJ_CASE_INSENSITIVE, NULL);
    
    status = NtOpenKey(&servicesKey, KEY_CREATE_SUB_KEY, &servicesOa);
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to open Services key: 0x%X\n", status);
        return FALSE;
    }
    
    // Create service subkey
    UNICODE_STRING serviceNameUs;
    InitUnicodeString(&serviceNameUs, serviceName);
    
    OBJECT_ATTRIBUTES serviceOa;
    InitObjectAttributes(&serviceOa, &serviceNameUs, OBJ_CASE_INSENSITIVE, servicesKey);
    
    status = NtCreateKey(
        &serviceKey,
        KEY_ALL_ACCESS,
        &serviceOa,
        0,
        NULL,
        REG_OPTION_NON_VOLATILE,
        &disposition
    );
    
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to create service key: 0x%X\n", status);
        goto cleanup;
    }
    
    wprintf(L"Service key created (disposition: %d)\n", disposition);
    
    // Set service parameters
    if (!SetRegistryDWord(serviceKey, L"Type", SERVICE_WIN32_OWN_PROCESS)) goto cleanup;
    if (!SetRegistryDWord(serviceKey, L"Start", SERVICE_AUTO_START)) goto cleanup;
    if (!SetRegistryDWord(serviceKey, L"ErrorControl", SERVICE_ERROR_NORMAL)) goto cleanup;
    if (!SetRegistryString(serviceKey, L"ImagePath", exePath)) goto cleanup;
    if (!SetRegistryString(serviceKey, L"DisplayName", displayName ? displayName : serviceName)) goto cleanup;
    if (!SetRegistryString(serviceKey, L"ObjectName", L"LocalSystem")) goto cleanup;
    
    if (description && wcslen(description) > 0) {
        SetRegistryString(serviceKey, L"Description", description);
    }
    
    wprintf(L"Service '%s' installed successfully via NT syscalls\n", serviceName);
    wprintf(L"Note: Service requires system reboot or manual SCM refresh to appear\n");
    
    success = TRUE;
    
cleanup:
    if (serviceKey) NtClose(serviceKey);
    if (servicesKey) NtClose(servicesKey);
    return success;
}

BOOL UninstallService(LPCWSTR serviceName) {
    HANDLE servicesKey = NULL;
    HANDLE serviceKey = NULL;
    NTSTATUS status;
    BOOL success = FALSE;
    
    if (!InitNtFunctions()) {
        wprintf(L"Failed to initialize NT functions\n");
        return FALSE;
    }
    
    // Try to stop service first (using ServiceController fallback)
    StopServiceByName(serviceName);
    
    // Open Services key
    UNICODE_STRING servicesPath;
    InitUnicodeString(&servicesPath, SERVICES_KEY_PATH);
    
    OBJECT_ATTRIBUTES servicesOa;
    InitObjectAttributes(&servicesOa, &servicesPath, OBJ_CASE_INSENSITIVE, NULL);
    
    status = NtOpenKey(&servicesKey, KEY_ENUMERATE_SUB_KEYS, &servicesOa);
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to open Services key: 0x%X\n", status);
        return FALSE;
    }
    
    // Open service subkey
    UNICODE_STRING serviceNameUs;
    InitUnicodeString(&serviceNameUs, serviceName);
    
    OBJECT_ATTRIBUTES serviceOa;
    InitObjectAttributes(&serviceOa, &serviceNameUs, OBJ_CASE_INSENSITIVE, servicesKey);
    
    status = NtOpenKey(&serviceKey, DELETE, &serviceOa);
    if (status != STATUS_SUCCESS) {
        if (status == STATUS_OBJECT_NAME_NOT_FOUND) {
            wprintf(L"Service '%s' does not exist\n", serviceName);
        } else {
            wprintf(L"Failed to open service key: 0x%X\n", status);
        }
        goto cleanup;
    }
    
    // Delete the key
    status = NtDeleteKey(serviceKey);
    if (status != STATUS_SUCCESS) {
        wprintf(L"Failed to delete service key: 0x%X\n", status);
        goto cleanup;
    }
    
    wprintf(L"Service '%s' uninstalled successfully via NT syscalls\n", serviceName);
    success = TRUE;
    
cleanup:
    if (serviceKey) NtClose(serviceKey);
    if (servicesKey) NtClose(servicesKey);
    return success;
}

BOOL StartServiceByName(LPCWSTR serviceName) {
    SC_HANDLE scManager = NULL;
    SC_HANDLE service = NULL;
    BOOL success = FALSE;
    
    scManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
    if (!scManager) {
        wprintf(L"OpenSCManager failed: %d\n", GetLastError());
        return FALSE;
    }
    
    service = OpenServiceW(scManager, serviceName, SERVICE_START | SERVICE_QUERY_STATUS);
    if (!service) {
        DWORD err = GetLastError();
        if (err == ERROR_SERVICE_DOES_NOT_EXIST) {
            wprintf(L"Service '%s' not found (may need reboot)\n", serviceName);
        } else {
            wprintf(L"OpenService failed: %d\n", err);
        }
        goto cleanup;
    }
    
    SERVICE_STATUS status;
    if (QueryServiceStatus(service, &status)) {
        if (status.dwCurrentState == SERVICE_RUNNING) {
            wprintf(L"Service '%s' is already running\n", serviceName);
            success = TRUE;
            goto cleanup;
        }
    }
    
    wprintf(L"Starting service '%s'...\n", serviceName);
    if (!StartServiceW(service, 0, NULL)) {
        wprintf(L"StartService failed: %d\n", GetLastError());
        goto cleanup;
    }
    
    // Wait for service to start
    Sleep(1000);
    if (QueryServiceStatus(service, &status)) {
        if (status.dwCurrentState == SERVICE_RUNNING) {
            wprintf(L"Service '%s' started successfully\n", serviceName);
            success = TRUE;
        } else {
            wprintf(L"Service state: %d\n", status.dwCurrentState);
        }
    }
    
cleanup:
    if (service) CloseServiceHandle(service);
    if (scManager) CloseServiceHandle(scManager);
    return success;
}

BOOL StopServiceByName(LPCWSTR serviceName) {
    SC_HANDLE scManager = NULL;
    SC_HANDLE service = NULL;
    BOOL success = FALSE;
    
    scManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
    if (!scManager) return FALSE;
    
    service = OpenServiceW(scManager, serviceName, SERVICE_STOP | SERVICE_QUERY_STATUS);
    if (!service) {
        CloseServiceHandle(scManager);
        return FALSE;
    }
    
    SERVICE_STATUS status;
    if (QueryServiceStatus(service, &status)) {
        if (status.dwCurrentState == SERVICE_STOPPED) {
            wprintf(L"Service '%s' is already stopped\n", serviceName);
            success = TRUE;
            goto cleanup;
        }
    }
    
    wprintf(L"Stopping service '%s'...\n", serviceName);
    if (!ControlService(service, SERVICE_CONTROL_STOP, &status)) {
        wprintf(L"ControlService failed: %d\n", GetLastError());
        goto cleanup;
    }
    
    // Wait for service to stop
    Sleep(1000);
    if (QueryServiceStatus(service, &status)) {
        if (status.dwCurrentState == SERVICE_STOPPED) {
            wprintf(L"Service '%s' stopped successfully\n", serviceName);
            success = TRUE;
        }
    }
    
cleanup:
    if (service) CloseServiceHandle(service);
    if (scManager) CloseServiceHandle(scManager);
    return success;
}

BOOL GetServiceStatusByName(LPCWSTR serviceName) {
    SC_HANDLE scManager = NULL;
    SC_HANDLE service = NULL;
    BOOL found = FALSE;
    
    scManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
    if (!scManager) {
        wprintf(L"OpenSCManager failed: %d\n", GetLastError());
        return FALSE;
    }
    
    service = OpenServiceW(scManager, serviceName, SERVICE_QUERY_STATUS | SERVICE_QUERY_CONFIG);
    if (!service) {
        DWORD err = GetLastError();
        if (err == ERROR_SERVICE_DOES_NOT_EXIST) {
            wprintf(L"Service '%s' does not exist in SCM\n", serviceName);
            wprintf(L"Note: Service may exist in registry but not yet loaded by SCM\n");
        } else {
            wprintf(L"OpenService failed: %d\n", err);
        }
        goto cleanup;
    }
    
    SERVICE_STATUS status;
    if (QueryServiceStatus(service, &status)) {
        wprintf(L"Service Name: %s\n", serviceName);
        wprintf(L"Status: ");
        switch (status.dwCurrentState) {
            case SERVICE_STOPPED: wprintf(L"Stopped\n"); break;
            case SERVICE_START_PENDING: wprintf(L"Start Pending\n"); break;
            case SERVICE_STOP_PENDING: wprintf(L"Stop Pending\n"); break;
            case SERVICE_RUNNING: wprintf(L"Running\n"); break;
            case SERVICE_CONTINUE_PENDING: wprintf(L"Continue Pending\n"); break;
            case SERVICE_PAUSE_PENDING: wprintf(L"Pause Pending\n"); break;
            case SERVICE_PAUSED: wprintf(L"Paused\n"); break;
            default: wprintf(L"Unknown (%d)\n", status.dwCurrentState);
        }
        wprintf(L"Service Type: Win32 Own Process\n");
        found = TRUE;
    }
    
cleanup:
    if (service) CloseServiceHandle(service);
    if (scManager) CloseServiceHandle(scManager);
    return found;
}
